### PR TITLE
Recipes App on Rails: V1

### PR DESCRIPTION
**Pull request by D4nielJ**
## Hello fellow code reviewer :wave: :vulcan_salute: :alien: 

This PR includes all of the features requested by Microverse, such as:
 
- A login page.
- A registration page.
- A food list (with all CRUD implementation, except for 'update').
- A recipes list (with all CRUD implementation, except for 'update').
- A general shopping list view (all missing food for all your recipes and total price).
- A list of all public recipes from other users with their names and total prices.

And the following requirements for groups of 2:

- You should follow the layout of the wireframes provided. You should personalize the rest of the design including colors, typographies, spacings, etc.
- Login page and registration page
- Food list
- Recipies list
- Public recipe list
- Recipe details
- General shopping list view
- Make sure there are no N+1 queries happening.
- Creates a navigation menu that allows users to open all of the pages you created.

Instructions on how to run this project can be found on the README file, besides, you can use the following emails and passwords to access the application:

```
leo@user.com
password

dan@user.com
password
```

### Thank you so much and happy coding! :tulip:

### CODE REVIEW

File: app/controllers/recipe_foods_controller.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
+class RecipeFoodsController < ApplicationController
+  before_action :authenticate_user!
+
+  def new
+    @recipe = Recipe.find(params[:recipe_id])
+    used = @recipe.food_used
+    @foods = Food.where(user_id: current_user.id).reject { |food| used.include?(food.name) }
```
Comment (oxenprogrammer) [inline app/controllers/recipe_foods_controller.rb]: **Optional Change**
- [x] It is good to keep the controller as lean as possible. This line could be a method in the Recipe Food model and just called in the controller.

File: app/controllers/shopping_controller.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  before_action :authenticate_user!
+
+  def index
+    @recipes = current_user.recipes
+
+    @foods = Food.where(user_id: current_user.id)
+
+    @total_price = 0
+    @recipes.each do |recipe|
+      @total_price += recipe.total_price
```
Comment (oxenprogrammer) [inline app/controllers/shopping_controller.rb]: **Optional Change**
- [x] It is good to keep the controller as lean as possible. This line could be a method/scope in the Shopping model and just called in the controller.

File: app/models/recipe.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+      total += ingredient.quantity * ingredient.food.price
+    end
+    total
+  end
+
+  def food_used()
+    used = Set.new
+    recipe_foods.includes(:food).each { |recipe_food| used.add(recipe_food.food.name) }
+    used
+  end
```
Comment (oxenprogrammer) [inline app/models/recipe.rb]: - [x] Good practice writing model based methods

File: app/views/shopping/_recipe_food_row.html.erb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<% qty = RecipeFood.joins(:food).where(food_id: food.id).sum('quantity') # There has to be a better way to do this %>
```
Comment (oxenprogrammer) [inline app/views/shopping/_recipe_food_row.html.erb]: **Optional Changes**
- [x] It is not good practice to write methods in the view. This belongs to the model or at worst to view helper methods. Let us consider refactoring it. You want to keep your view as dumb as possible. Remeber `KISS` principle?

File: app/views/shopping/_recipe_food_row.html.erb [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+<% if qty > 0 %>
+  <tr>
+    <td><%= food.name %></td>
+    <td><%= qty %></td>
+    <td><%= food.price * qty %></td>
+  </tr>
+<% end %>
```
Comment (oxenprogrammer) [inline app/views/shopping/_recipe_food_row.html.erb:8]: **Optional Changes**
- [ ] Logic belongs to the view helper file. Again, you want to follow the `KISS` principle here.

**Review by oxenprogrammer — APPROVED** [review]
Hi @D4nielJ and @leonardodiasb, 

In my opinion, your project is ready for the presentation!  There is nothing else to say other than... it's time to merge it :shipit: 


Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

File: app/views/shopping/_recipe_food_row.html.erb (hunk 5 above)
Comment (D4nielJ) [reply app/views/shopping/_recipe_food_row.html.erb:8]: I'm not sure about how to better refactor this. It's conditional rendering, which seems appropriate in a view. I could go with a view helper, but the fact that I'm moving such a small condition to a helper method it's not very `KISS` in my opinion.

**Pull request merged**